torch/csrc/jit/runtime/graph_iterator.h

```diff
@@ -5,113 +5,196 @@ namespace jit {
-          // else_block to the stack so that it will be processed after.
-          block_stack_.emplace_back(else_block, else_block->nodes().begin());
-          current_.first = then_block;
-          current_.second = then_block->nodes().begin();
-        }
-      } else if (node->kind() == prim::Loop || node->kind() == prim::With) {
-        auto* body_block = node->blocks().at(0);
+  // Moves to the next adjacent node or up in to the parent if that is not
+  // possible.
+  void move_next() {
+    auto block = current_->owningBlock();
+    Node* previous = *current_;
+
+    // Increment to the next node in the current block.
+    ++current_;
 
-        bool body_block_empty =
-            body_block->nodes().begin() == body_block->nodes().end();
+    // Check if we're at the end of the block. If so we need
+    // to move upwards (if it makes sense to).
+    if (current_ == block->nodes().end()) {
+      move_up(previous);
+    }
+  }
 
-        if (!body_block_empty) {
-          // If body_block is not empty, push the current block onto the stack
-          // to resume processing it later and set current_ to {body_block,
-          // body_block.begin()}.
-          block_stack_.emplace_back(current_.first, current_.second);
+  // Moves to the next node in the graph into children if it can.
+  void move_into() {
+    if (*current_ == nullptr) {
+      return;
+    }
 
-          current_.first = body_block;
-          current_.second = body_block->nodes().begin();
-        }
+    auto node = *current_;
+
+    // Check if we're currently on a node that contains sub-nodes.
+    if (node->kind() == prim::If) {
+      // If nodes can have two child blocks. From an If node
+      // in `move_into` we only enter the first block which contains
+      // elements. After processing of those elements is done, we move
+      // to the next block (the else block if it has nodes) from the
+      // `move_up` call.
+      auto* then_block = node->blocks().at(0);
+      auto* else_block = node->blocks().at(1);
+
+      bool then_block_empty =
+          then_block->nodes().begin() == then_block->nodes().end();
+
+      bool else_block_empty =
+          else_block->nodes().begin() == else_block->nodes().end();
+
+      if (!then_block_empty) {
+        current_ = then_block->nodes().begin();
+      } else if (!else_block_empty) {
+        current_ = else_block->nodes().begin();
+      } else {
+        // This `if` block does not have any child nodes so we need to continue
+        // moving next/over it.
+        move_next();
       }
-    } else {
-      // There are no more nodes in the current block. Resume processing of the
-      // block on the top of the stack if there is one.
-      if (!block_stack_.empty()) {
-        current_ = block_stack_.back();
-        block_stack_.pop_back();
+    } else if (node->kind() == prim::Loop || node->kind() == prim::With) {
+      auto* body_block = node->blocks().at(0);
```

Comment:
would it work to set current_ to `node->blocks()->param_node()` and then call move_next() ? same question would apply for prim::If handling within this function as well

Reply:
`param_node` doesn't have a next node. Per the docs in `OVERVIEW.md`

> The prim::Param node represents the inputs to the Block and does not have a prev() or next() Node.